### PR TITLE
Add recipe for branch-based secrets management

### PR DESCRIPTION
**Pull request by c-w**
This pull request adds a section about branch-based secrets management to the engineering playbook. The section is based on findings from a customer project and answers questions such as:

- Why do we need branch-based secrets management?
- How can we implement branch-based secrets management?

The section is included as part of the continuous deployment heading since it already has some prior content around secrets management.

### CODE REVIEW

**Comment by fnocera** [comment]
This is great!

**Comment by fnocera** [comment]
@SaraSp @cloudbeatsch flagging for review, thanks!

**Comment by charleszipp** [comment]
I would invite @kevinhartman to take a look at the secrets part with node. We recently opened a story in our current project because in ACI, env vars are visible in the portal (thats how i understood what @kevinhartman and narmatha reporting). So is the env the best place to put secrets? Is there a different place for node containers that we should place secrets?

**Comment by fnocera** [comment]
@charleszipp I will message @kevinhartman to add a recipe for node secrets!

**Comment by c-w** [comment]
@charleszipp The recommendation for using dotenv in Node came from @tobecodex so perhaps he can add more context.

Personally, I like the way the section on [Enhanced-Security Applications](https://github.com/microsoft/code-with-engineering-playbook/blob/ea6bd9beb9af4af2996d012c93c36111d13e288b/continuous-deployment/secrets-management/readme.md#enhanced-security-applications) puts it: using environment variables is fine for most applications unless you have specific additional security requirements (in which case you'd likely have a security review, pentesting, etc. all set up which will inform the best enhanced security strategy for your use-case).

File: continuous-deployment/secrets-management/recipes/azure-devops/secrets-per-branch.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+A solution to these limitations is to manage secrets in the git repository jointly with the project's source code. As described in [secrets management](../../readme.md), secrets should never be checked into the repository in plain text, however, we can add an encrypted version of our secrets to the repository and enable our CI/CD agents and developers to decrypt the secrets for local usage with some pre-shared key. This gives us the best of both worlds: a secure storage for secrets as well as side-by-side management of secrets and code.
+
+```sh
+# first, make sure that we never commit our plain text secrets and generate a strong encryption key
+echo ".env" >> .gitignore
+ENCRYPTION_KEY="$(< /dev/urandom tr -dc '_A-Z-a-z-0-9' | head -c128)"
```
Comment (davidkpiano) [inline continuous-deployment/secrets-management/recipes/azure-devops/secrets-per-branch.md]: This produces `Illegal byte sequence` on MacOS

File: continuous-deployment/secrets-management/recipes/azure-devops/secrets-per-branch.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+MY_SECRET=
+"
+
+# next, encrypt the plain text secrets; the resulting .env.enc file can safely be committed to the repository
+echo "${ENCRYPTION_KEY}" | openssl enc -aes-256-cbc -md sha512 -pass stdin -in .env -out .env.enc
+git add .env.enc .env.template
+git commit -m "Update secrets"
+```
+
+When running the CI/CD, the build server can now easily access the secrets by decrypting them. E.g. for Azure Devops, configure `ENCRYPTION_KEY` as a [secret pipeline variable](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables#secret-variables) and then add the following step to `azure-pipelines.yml`:
```
Comment (davidkpiano) [inline continuous-deployment/secrets-management/recipes/azure-devops/secrets-per-branch.md]:
```suggestion
When running the CI/CD, the build server can now easily access the secrets by decrypting them. E.g. for Azure DevOps, configure `ENCRYPTION_KEY` as a [secret pipeline variable](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables#secret-variables) and then add the following step to `azure-pipelines.yml`:
```

File: continuous-deployment/secrets-management/readme.md [hunk 3]
```diff
@@ -17,6 +17,7 @@ In a production system, assuming Azure, the secrets would be created in the envi
 the Azure CLI to do the same is a useful time-saving utility. See [here](https://docs.microsoft.com/en-us/cli/azure/webapp/config/appsettings?view=azure-cli-latest) for more details.
 
 *It is best practice to maintain separate secrets configurations for each separate environment that you run. e.g. dev, test, prod, local etc*
+When using a branch-based deployment strategy (e.g. 'master' deploys to production, 'development' deploys to staging, and so forth), a simple way to manage separate secrets configurations on a per environment basis is described in the [secrets-per-branch recipe](./recipes/azure-devops/secrets-per-branch.md).
```
Comment (davidkpiano) [inline continuous-deployment/secrets-management/readme.md]:
```suggestion
When using a branch-based deployment strategy (e.g. 'master' deploys to production, 'development' deploys to staging, and so forth), a simple way to manage separate secrets configurations on a per-environment basis is described in the [secrets-per-branch recipe](./recipes/azure-devops/secrets-per-branch.md).
```

File: continuous-deployment/secrets-management/recipes/azure-devops/secrets-per-branch.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+- *Pipeline variables are global shared state.* This can easily lead to confusing situations and hard to debug problems when multiple developers make concurrent changes to the pipeline variables which may override each other. Having a single global set of pipeline variables also makes it impossible for secrets to vary per environment (e.g. when using a branch-based deployment model where 'master' deploys using the production secrets, 'development' deploys using the staging secrets, and so forth).
+
+A solution to these limitations is to manage secrets in the git repository jointly with the project's source code. As described in [secrets management](../../readme.md), secrets should never be checked into the repository in plain text, however, we can add an encrypted version of our secrets to the repository and enable our CI/CD agents and developers to decrypt the secrets for local usage with some pre-shared key. This gives us the best of both worlds: a secure storage for secrets as well as side-by-side management of secrets and code.
```
Comment (davidkpiano) [inline continuous-deployment/secrets-management/recipes/azure-devops/secrets-per-branch.md]:
```suggestion
A solution to these limitations is to manage secrets in the Git repository jointly with the project's source code. As described in [secrets management](../../readme.md), secrets should never be checked into the repository in plain text, however, we can add an encrypted version of our secrets to the repository and enable our CI/CD agents and developers to decrypt the secrets for local usage with some pre-shared key. This gives us the best of both worlds: a secure storage for secrets as well as side-by-side management of secrets and code.
```

File: continuous-deployment/secrets-management/recipes/azure-devops/secrets-per-branch.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+```sh
+# first, make sure that we never commit our plain text secrets and generate a strong encryption key
+echo ".env" >> .gitignore
+ENCRYPTION_KEY="$(< /dev/urandom tr -dc '_A-Z-a-z-0-9' | head -c128)"
+
+# now let's add some secret to our dotenv file
```
Comment (davidkpiano) [inline continuous-deployment/secrets-management/recipes/azure-devops/secrets-per-branch.md]:
```suggestion
# now let's add some secret to our .env file
```

File: continuous-deployment/secrets-management/recipes/azure-devops/secrets-per-branch.md (hunk 1 above)
Comment (c-w) [reply continuous-deployment/secrets-management/recipes/azure-devops/secrets-per-branch.md]: Should be fixed by 25fa945

**Pull request merged**